# Estatística & Probabilidade - 2023/02

## Projeto 01 - Calculando probabilidades com simulação estocástica

### Professor: Hugo Tremonte de Carvalho

#### Manoel Marcelo da Silva - 121088349

#### Vinícius Lima - 1210xxxxx


# ⚠️ INSTRUÇÕES ⚠️ LEIAM ATENTAMENTE! ⚠️

* A data limite de entrega da avaliação é sábado 01/10/2023 às 23h59. Avaliações entregues após esse prazo serão desconsideradas.

* A entrega deve ser feita exclusivamente através do Google Classroom, usando a linguagem Python dentro de um notebook do Google Colaboratory, a ser anexado na atividade da seguinte forma:
  - Dentro da turma no Google Classroom, clique em "Atividades"
            
  - No tema "Projetos", clique em "Projeto 02" e em seguida em "Ver atividade"
            
  - No canto superior direito, clique em "+ Adicionar ou criar" e em seguida selecione em "Arquivo" ou "Google Drive", se optar fazer upload do notebook ou selecionar um arquivo de seu Google Drive, respectivamente.
            
  - O título do notebook deverá ser, obrigatoriamente, "SEU NOME - Projeto 01".
            
  - Para entregar o projeto, abra a atividade que o contém (conforme procedimento explicado no ponto acima), e no canto superior direito clique em "ENTREGAR" (o texto estará em uma caixa cinza, mas ele é clicável).
            
  - Na tela a seguir clique em "Entregar" para confirmar a entrega do trabalho.
            
  - Se você se arrependeu e quiser fazer outra entrega, basta abrir a atividade, no canto superior direito clique em "Cancelar envio" e confirme clicando novamente em "Cancelar envio" na janela que irá abrir. Você pode fazer isso quantas vezes quiser até o prazo determinado.
            
  - As explicações que você precisará fazer devem ser feitas em células de texto em seu notebook. Somente comentários pequenos devem ser feitos no código.
            
  - Caso seja necessário, equações podem ser introduzidas em $\LaTeX$ nas células de texto.
            
  - Figuras também podem ser incluídas em células de texto, tanto através de upload quanto simplesmente arrastando-a para alguma célula de texto. Nesse caso, sempre informe a fonte de onde a figura foi retirada, caso não tenha sido gerada por você.

* O projeto poderá ser feito **até** em dupla: quem quer fazer sozinho/a faz, mas quem quiser fazer em dupla também tem esse direito. Porém, vamos evitar problemas de "eu fiz tudo e ele/a não fez nada"! Infelizmente não tenho como controlar isso, e após o projeto entregue não adianta "eu fiz tudo sozinho/a", pois eu não estava lá para ver! Tentem se resolver entre si, mas em último caso posso auxiliar com uma mediação.

* No caso de trabalho em dupla, apenas **uma** pessoa da dupla deverá fazer o envio do trabalho, claro, deixando explícito quem é o/a parceiro/a.

* A troca de conhecimento na realização da avaliação é permitida e encorajada: ciência se faz com colaboração, e devemos seguir esse espírito aqui. Porém, cada dupla (ou indivíduo) deverá ter a sua própria resolução, e cópias ou outras ilegalidades serão severamente punidas com a anulação da avaliação para o(s) aluno(s) suspeito(s).

* Todos os passos de sua resolução devem ser devidamente justificados.

* Ao entregar essa avaliação, você afirma ter lido e estar de acordo com essas regras, comprometendo-se a cumpri-las.

# O projeto

O objetivo da atividade é que vocês calculem algumas probabilidades através de simulação estocástica, ou seja, através de um grande número de sorteios aleatórios devidamente projetados para calcular a probabilidade desejada. Algumas dessas probabilidades podem ser calculadas analiticamente com certa facilidade, mas em outros pode ser **bem** difícil (ou mesmo impossível). Dessa forma, o cálculo de probabilidades através de simulação estocástica pode ser tanto uma ferramenta fundamental para resolver algum problema ou para te dar alguma intuição, te guiando para o resultado analítico. Em alguns momentos vocês deverão justficar teoricamente os resultados obtidos, pois senão, não tem graça! 😁

## Questão 1: Calculando $\pi$ jogando dardos

#### a) Suponha que num jogo de dardos você tem, pendurado em uma parede, o alvo (círculo de raio 1) devidamente emoldurado (inscrito em um quadrado de lado 2 - as unidades não importam aqui). Porém, suas habilidades ao arremessar um dardo em direção ao alvo são péssimas, e você pode atingir qualquer ponto dentro do quadrado uniformemente (note que, forças ocultas e inexplicáveis do Universo fazem com que seja impossível você acertar *fora* da moldura). Através de um cálculo analítico mostre como que a probabilidade de você acertar o alvo se relaciona com o número $\pi$.

Considerando a probabilidade de acertar uniformemente qualquer ponto dentro do quadrado, a probabilidade de acertar o alvo é a razão entre a área do alvo e a área do quadrado. A área do alvo é $\pi r^2$ e a área do quadrado é $(2r)^2$, então:

$$\mathbb{P} = \frac{\text{Eventos de interesse}}{\text{Espaço amostral}}$$
$$\mathbb{P} = \frac{\text{Área do círculo}}{\text{Área do quadrado}}$$
$$\mathbb{P} = \frac{\pi r^2}{(2r)^2}$$
$$\mathbb{P} = \frac{\pi}{4}$$
$$\pi = 4\mathbb{P}$$

Portanto, $\pi$ é quatro vezes a probabilidade de acertar o alvo.

#### b) Utilizando geração de números aleatórios, faça uma simulação para estimar o valor de $\pi$ através do procedimento acima. Ou seja, simule que dardos são arremessados uniformemente no quadrado e utilize a interpretação frequentista da probabilidade para aproximar o valor de $\pi$. Note que você precisará de "muitas" tentativas para ter uma boa estimativa. O quão grande você acha que "muitas" deve ser?

In [1]:
import random
import numpy as np

In [2]:
def simulacao_pi(n):
    
    dentro = 0
    fora = 0
    
    for _ in range(n):
        x = random.random() #Escolhe um número uniformemente aleatório entre 0 e 1 para x no primeiro quadrante
        y = random.random() #Escolhe um número uniformemente aleatório entre 0 e 1 para y no primeiro quadrante
        
        if (x**2 + y**2) > 1: # Usa desigualdade triangular para identificar se o ponto está fora do círculo
            fora += 1
        else:
            dentro += 1
    
    print(f'Pontos dentro: {dentro}')
    print(f'Pontos fora: {fora}')
    print(f'Total: {dentro + fora}')
    
    return float(4 * dentro / (dentro + fora))


Mesmo princípio do código acima, porém com execução mais rápida.

In [3]:
def simulacao_pi_rapido(n):
    
    """Simula n pontos aleatóios no quadrado de lado 2 e retorna a aproximação de pi."""
    
    x = np.random.uniform(0, 1, n) #Escolhe n números aleatórios entre 0 e 1 para x no primeiro quadrante
    y = np.random.uniform(0, 1, n) #Escolhe n números aleatórios entre 0 e 1 para y no primeiro quadrante
    
    return (4 * np.mean(x**2 + y**2 <= 1)) #Multiplica por 4 a média dos pontos dentro do círculo

In [4]:
print("Erro da simulação com 100 de pontos: ", abs(np.pi - simulacao_pi_rapido(100)))
print("Erro da simulação com 10000 de pontos: ", abs(np.pi - simulacao_pi_rapido(10000)))
print("Erro da simulação com 100000000 de pontos: ", abs(np.pi - simulacao_pi_rapido(100000000)))
#print("Erro da simulação com 1000000000 de pontos: ", abs(np.pi - simulacao_pi_rapido(1000000000)))



Erro da simulação com 100 de pontos:  0.2615926535897932
Erro da simulação com 10000 de pontos:  0.008792653589793087
Erro da simulação com 100000000 de pontos:  0.0003710664102070105


A aproximação de $\pi$ não é ótima, pois requer muitas iterações para se aproximar de $\pi$ com um erro relativamente alto. Percebemos de forma empírica que o erro reduz conforme o número de iterações aumenta, isso faz sentido, pois quanto mais pontos gerados (tentendo ao infinito), mais próximo do valor real de $\pi$ estaremos. Contudo, há a questão da eficiência, pois quanto mais pontos gerados, mais tempo o algoritmo leva para executar. Além disso, o resultado é impactado pela randomicidade dos pontos gerados, o que influencia no erro.

Assim, o número de iterações deve ser grande o suficiente para que o erro seja aceitável, mas não tão grande a ponto de tornar o algoritmo ineficiente. Nesse contexto, o maior número viável para se computar é o $n = 100000000$ ou $10^8$, o que dá uma aproximação de $\pi$ com erro médio de $0.0001$ ou $10^{-4}$. Acima disso, o tempo de execução é inviável.

## Questão 2: O problema de Monty Hall

Em um famoso programa de televisão da década de 70 nos Estados Unidos, Monty Hall, o apresentador, apresentava três portas aos concorrentes. Atrás de uma delas estava um prêmio e, atrás das outras duas não havia nada. A interação se dava em três etapas:

* Na 1a. etapa o concorrente escolhe uma das três portas (que ainda não é aberta);
* Na 2a. etapa, Monty abre uma das outras duas portas que o concorrente não escolheu, revelando que o prêmio não se encontra nessa porta e revelando uma porta vazia (note que o apresentador **sempre** irá abrir uma porta que está vazia, e **nunca** a porta com o prêmio);
* Na 3a. etapa Monty pergunta ao concorrente se quer decidir permanecer com a porta que escolheu no início do jogo ou se ele pretende mudar para a outra porta que ainda está fechada para então a abrir. Agora, com duas portas apenas para escolher — pois uma delas já se viu, na 2a. etapa, que não tinha o prêmio — e sabendo que o prêmio está atrás de uma das restantes duas, o concorrente tem que tomar a decisão.

O objetivo dessa questão é estudar qual é a estratégia mais lógica: ficar com a porta escolhida inicialmente ou mudar de porta? Do ponto de vista probablístico, com qual das duas portas ainda fechadas o concorrente tem mais probabilidades de ganhar?

#### a) Primeiramente, pensando do ponto de vista do participante, argumente se é mais razoável que ele tome sua decisão pensando do ponto de vista da interpretação frequentista ou subjetiva do conceito de probabilidade. Claro, justifique sua escolha!

Obs.: Esse item é engraçado pois uma vez que o cálculo da probabilidade independe da interpretação escolhida, qualquer resposta é válida, do ponto de vista formal. Porém, a graça aqui é *como* você vai fazer a sua justificativa para a interpretação escolhida!

Assumir uma interpretação frequentista nesse problema pode ser vantajoso no caso do apresentador ser enviezado, ou seja, não possuir uma distribuição uniforme na posição da porta do prêmio. Nesse caso, o apresentador pode ter uma preferência por uma porta específica, o que pode ser explorado pelo participante. Além disso, é possível perceber se a possibilidade de trocar de porta é mais vantajosa analisando o resultado de participantes anteriores. Assim, a partir de dados anteriores, é possível encontrar um padrão em que trocar a porta é mais vantajoso ou não.

Contudo, caso não haja muitos dados ou o apresentador é muito imprevisível, a interpretação subjetiva é mais vantajosa, pois o participante pode escolher sempre maximizar sua probabilidade de ganhar, independente do apresentador ou de dados anteriores a partir da intuição de que trocar a porta sempre é vantagem (o que no caso é verdade, estaríamos ganhando 33% a mais de chance de ganhar). Nesse caso, o participante pode escolher sempre trocar a porta, pois sua probabilidade de ganhar é maior.

Ambas as interpretações são válidas dependendo do contexto em que o problema está inserido.


#### b) Talvez você já saiba a resposta para esse problema, uma vez que ele é bem famoso, mas não tem problema! Faça simulações para encontrar qual é o melhor cenário para o participante: manter sua escolha inicial ou mudar de porta. Justifique, **intuitivamente** (ou seja, **sem fazer contas**) porque essa escolha é razoável.

In [5]:
def monty_hall(n, p):
    
    """Simula n vezes o jogo Monty Hall um número p de portas e retorna a probabilidade de ganhar se trocar de porta."""
    
    ganhou = 0
    
    for _ in range(n):
        porta = np.random.randint(1, p+1) #Escolhe uma porta aleatória
        premio = np.random.randint(1, p+1) #Escolhe uma porta aleatória para o prêmio
        
        if porta != premio: #Se a porta escolhida for diferente da porta do prêmio
            ganhou += 1 #Troca de porta e ganha o prêmio
    
    return ganhou / n #Retorna a probabilidade de ganhar

In [6]:
print("Probabilidade de ganhar se trocar de porta com 3 portas: ", monty_hall(1000000, 3))
print("Probabilidade de ganhar se trocar de porta com 4 portas: ", monty_hall(1000000, 4))
print("Probabilidade de ganhar se trocar de porta com 999 portas: ", monty_hall(1000000, 999))

Probabilidade de ganhar se trocar de porta com 3 portas:  0.66594
Probabilidade de ganhar se trocar de porta com 4 portas:  0.750597
Probabilidade de ganhar se trocar de porta com 999 portas:  0.99902


O problema de Monty Hall é um problema clássico de probabilidade, em que nesse jogo a probalilidade de ganhar é maior se o participante escolher trocar de porta, apesar de não ser muito intuitivo (já que a maioria das pessoas assumem que a probabilidade é 50% erroneamente) é mais fácil analisar esse problema mudando o número de portas para um valor extremente grande.

Vamos assumir que temos agora 1000 portas para escolher uma delas, ao decorrer do jogo, o apresentador vai remover todas as outras portas vazias e deixar apenas duas portas, a porta escolhida pelo participante e outra porta. Nesse caso, a probabilidade de ganhar é de 99.9% se o participante escolher trocar de porta, pois a probabilidade de escolher a porta com o prêmio é de 1/1000 inicialmente, enquanto a probabilidade das portas restantes de ter o prêmio aumentam cada vez que o apresentador remove uma porta. Isso acontece porque o apresentador sempre remove uma porta vazia, nunca uma com o prêmio. Assim, a probabilidade de ganhar se o participante escolher trocar de porta é de 999/1000, enquanto a probabilidade de ganhar se o participante escolher manter a porta é de 1/1000.

O mesmo princípio existe com 3 portas. Imagine 2 grupos, um com apenas 1 porta que o participante escolheu inicialmente e outro com o restante das portas. A "porcentagem" da porta que o apresentador remove ainda pertence ao grupo em que ela reside, assim, a probabilidade de ganhar se o participante escolher trocar de porta é de 2/3, enquanto a probabilidade de ganhar se o participante escolher manter a porta é de 1/3. Sendo mostrado acima pela simulação e logo abaixo pela solução analítica usando probabilidade condicional.

#### c) Agora sim, justifique fazendo as contas a probabilidade encontrada no item b). Caso você pesquise por aí, você vai encontrar algumas soluções utilizando o Teorema de Bayes. Como nós não vimos ainda tal conteúdo, tente fazer somente com probabilidade condicional. Juro que é possível, e vai ser legal para treinar o conceito recém-aprendido.

Considere três eventos:

$A$ = Evento em que o prêmio está na porta B

$B$ = Evento em que o participante escolhe a porta B

$C$ = Evento em que o apresentador abre a porta A

Para o participante ganhar o prêmio em que esteja na primeira porta escolhida, ele deve escolher a porta B e o apresentador abrir a porta A. Portanto, a probabilidade de ganhar o prêmio em que esteja na primeira porta escolhida é:

$$\mathbb{P}(A|BC) = \frac{\mathbb{P}(ABC)}{\mathbb{P}(BC)}$$

Decompondo $ \mathbb{P}(ABC) $ em $ \mathbb{P}(C|AB)*\mathbb{P}(AB) $ e $ \mathbb{P}(BC) $ em $ \mathbb{P}(C|B)*\mathbb{P}(B) $ pela definição de probabilidade condicional e pela regra da multiplicação, temos:

$$\mathbb{P}(A|BC) = \frac{\mathbb{P}(ABC)}{\mathbb{P}(B)*\mathbb{P}(C|B)}$$
$$\mathbb{P}(A|BC) = \frac{\mathbb{P}(C|AB)*\mathbb{P}(AB)}{\mathbb{P}(B)*\mathbb{P}(C|B)}$$
$$\mathbb{P}(A|BC) = \frac{\mathbb{P}(C|AB)*\mathbb{P}(B|A)*\mathbb{P}(A)}{\mathbb{P}(B)*\mathbb{P}(C|B)}$$

Como o apresentador sempre abre uma porta sem prêmio, então $\mathbb{P}(C|AB) = 1/2$ e $\mathbb{P}(C|B) = 1/2$, pois há duas portas livres para ele escolher livremente.
Além disso, $\mathbb{P}(B|A) = 1/3$, $\mathbb{P}(B) = 1/3$ e $\mathbb{P}(A) = 1/3$. Portanto:

$$\mathbb{P}(A|BC) = \frac{1/2*1/3*1/3}{1/3*1/2}$$
$$\mathbb{P}(A|BC) = \frac{1}{3}$$

Assim, seu complementar é o evento em que o prêmio não está na primeira porta escolhida, ou seja, está na segunda ou terceira porta. Portanto, a probabilidade de ganhar o prêmio em que esteja na segunda ou terceira porta é: 
$$\mathbb{P}(A^{c}|BC) = 1 - \mathbb{P}(A|BC)$$
$$\mathbb{P}(A^{c}|BC) = \frac{2}{3}$$


## Questão 3: Jogo da velha aleatório

Agora, vamos tentar estimar uma probabilidade cujo cálculo analítico é bem mais complexo. Suponha que dois jogadores joguem jogo da velha porém com uma estratégia um pouco exótica: cada jogador escolhe a casa onde irá marcar $\times$ ou $\circ$ uniformemente dentre as casas vazias. Exemplificando, o primeiro jogador marca $\times$ em uma das 9 casas com igual probabilidade; o segundo marca $\circ$ em uma das 8 casas restantes com igual probabilidade; e assim sucessivamente até algum jogador ganhar (linha, coluna ou diagonal completa com $\times$ ou $\circ$) ou empatar (todas as células preenchidas porém sem nenhuma linha nem coluna nem diagonal completa com $\times$ ou $\circ$).

#### a) Encontre uma aproximação para a probabilidade do jogo **não** terminar em empate. Tente encontrar uma justificativa intuitiva para sua resposta.

Para encontrar a probabilidade do jogo não terminar em empate, buscamos determinar as chances de vitória do primeiro jogador **ou** do segundo jogador. Para isso, simulamos uma grande quantidade de jogos da velha a fim de contabilizar as vitórias do primeiro jogador, do segundo jogador e dos empates. Nosso código conta com os seguintes passos:


- Dado um tabuleiro vazio de jogo da velha, escolhemos aleatoriamente uma casa dentre as $9$ disponíveis a cada rodada para preenchê-la com "X" (dá inicío à primeira rodada) e "O" sucessivamente.

- Sendo completada uma combinação de vitória para o jogador "X" ou "O", o jogo termina e uma vitória é acrescentada na contagem `vitoriasX` ou `vitoriasO`.

- Após $9$ rodadas (máximo possível tendo em vista o número de casas do tabuleiro), não havendo vitória para nenhum dos jogadores, o jogo termina empatado e, portanto, um empate é contabilizado na contagem `empates`.

- Ao final de um número considerável de partidas (previamente determinado em `numJogos`), é possível verificar o valor das probabilidades de vitórias e empates, os quais são obtidos conforme a fórmula: $\mathbb{P} = \frac{\text{Eventos de interesse}}{\text{Espaço amostral}}$ . Logo, as probabilidades de vitória do jogador "X", do jogador "O" e de empates são dadas respectivamente por `vitoriasX/numJogos`, `vitoriasO/numJogos` e `empates/numJogos`.


Para verificarmos uma vitória de um dos jogadores, é necessário que, a cada rodada, pelo menos uma das seguintes combinações ocorra nas casas do tabuleiro ao serem preenchidas com "X" ou "O":

1. Uma das três linhas na horizontal completamente preenchidas ($3$ combinações);

2. Uma das três colunas na vertical completamente preechidas ($3$ combinações);

3. Uma das duas diagonais completamente preenchidas ($2$ combinações).


Assim, sendo inicialmente $8$ o total de combinações de vitória para cada jogador, podemos simular os jogos da velha com o seguinte código:

In [7]:
def jogoDaVelha(numCasas):
  tabuleiro = np.zeros(numCasas)
  rodadaX = True

  for rodada in range(numCasas):
    # Seleciona uma casa aleatoriamente
    casa = random.randint(0, numCasas - 1)

    # Busca uma casa ainda não preenchida
    while(tabuleiro[casa] > 0):
      casa = random.randint(0, numCasas - 1)

    # Marca uma casa com "X" ou "O" dependendo da rodada
    if(rodadaX):
      tabuleiro[casa] = 1
    else:
      tabuleiro[casa] = 2

    # Altera entre a rodada "X" e "O"
    rodadaX = not rodadaX

    # Para a função se o jogo tiver terminado
    if numCasas == 27:
      resultado = verificaResultadoTridimencional(tabuleiro, rodada)
    else:
      resultado = verificaResultado(tabuleiro, rodada)

    if resultado >= 0:
      break

  return resultado


def verificaResultado(tabuleiro, rodada):
  # Casos de vitória "X"
  if tabuleiro[0] == 1 and tabuleiro[1] == 1 and tabuleiro[2] == 1: return 1
  elif tabuleiro[3] == 1 and tabuleiro[4] == 1 and tabuleiro[5] == 1: return 1
  elif tabuleiro[6] == 1 and tabuleiro[7] == 1 and tabuleiro[8] == 1: return 1

  elif tabuleiro[0] == 1 and tabuleiro[3] == 1 and tabuleiro[6] == 1: return 1
  elif tabuleiro[1] == 1 and tabuleiro[4] == 1 and tabuleiro[7] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[5] == 1 and tabuleiro[8] == 1: return 1

  elif tabuleiro[0] == 1 and tabuleiro[4] == 1 and tabuleiro[8] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[4] == 1 and tabuleiro[6] == 1: return 1

  # Casos de vitória "O"
  elif tabuleiro[0] == 2 and tabuleiro[1] == 2 and tabuleiro[2] == 2: return 2
  elif tabuleiro[3] == 2 and tabuleiro[4] == 2 and tabuleiro[5] == 2: return 2
  elif tabuleiro[6] == 2 and tabuleiro[7] == 2 and tabuleiro[8] == 2: return 2

  elif tabuleiro[0] == 2 and tabuleiro[3] == 2 and tabuleiro[6] == 2: return 2
  elif tabuleiro[1] == 2 and tabuleiro[4] == 2 and tabuleiro[7] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[5] == 2 and tabuleiro[8] == 2: return 2

  elif tabuleiro[0] == 2 and tabuleiro[4] == 2 and tabuleiro[8] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[4] == 2 and tabuleiro[6] == 2: return 2

  else:
    # Verifica se é a última rodada e determina o empate
    if rodada == 8: return 0
    else: return -1


# Simulando os jogos
numJogos = 100000
vitoriasX = 0.
vitoriasO = 0.
empates = 0.

for i in range(numJogos):
  resultado = jogoDaVelha(9)
  if resultado == 0: empates += 1
  elif resultado == 1: vitoriasX += 1
  elif resultado == 2: vitoriasO += 1

# Imprimindo resultados
print(f'Vitórias X: {round((vitoriasX / numJogos) * 100, 2)}%')
print(f'Vitórias O: {round((vitoriasO / numJogos) * 100, 2)}%')
print(f'Empates: {round((empates / numJogos) * 100, 2)}%')

Vitórias X: 58.36%
Vitórias O: 28.93%
Empates: 12.71%


Portanto, dadas as probabilidades acima considerando uma simulação de $100000$ jogos, podemos concluir que a chance de um jogo não terminar em empate é de aproximadamente:

$\text{vitóriasX + vitoriasO} = 58,59\% + 28,84\% = 87,43\%$

Tal resultado se mostra consistente, já que intuitivamente há menos possibilidades de ocorrer empate do que vitória para um dos jogadores, uma vez que, para que haja empate, é necessário que todas as casas do tabuleiro sejam preenchidas sem que nenhuma das combinações de vitória ocorra, e isso só pode ser possível de acontecer em 3 tipos de posições no tabuleiro apenas. 

Por outro lado, para que haja vitória de um dos jogadores, basta que uma das combinações de vitória ocorra, o que é mais provável de acontecer do que o empate pois há muito mais combinações de posições para um dos jogadores ganhar. 

Isso também acontece pois estamos analisando um jogo com posições aleatórias, ou seja, não há uma estratégia definida para que um dos jogadores vença, o que torna o jogo mais imprevisível e, portanto, mais provável de terminar em vitória para um dos jogadores do que em empate.

#### b) Dentre os jogos que não terminam empatados, qual jogador tem mais chances de ganhar, se ambos utilizam essa estratégia: o jogador que começa ou o segundo jogador? Tente encontrar uma justificativa intuitiva para sua resposta.

Ao analisarmos as probabilidades obtidas no item **(a)**, considerando a aleatoriade das jogadas, é possível obter facilmente o jogador com mais chances de ganhar dentre os jogos que não terminam empatados. Para isso, basta manteremos o número de vitórias dos jogadores "X" e "O" como *eventos de interesse* e limitar o *espaço amostral* a somente os jogos que não terminaram empatados (dados por `vitoriasX + vitoriasO`).

Assim, podemos ver pelo seguinte código que o primeiro jogador, isto é o jogador "X", é o que tem mais chances de ganhar dentre os jogos não empatados:

In [8]:
jogosNaoEmpatados = vitoriasX + vitoriasO

# Imprimindo resultados
print('----- Jogos que não terminam empatados -----\n')
print(f'Chance do jogador "X" ganhar: {round((vitoriasX / jogosNaoEmpatados) * 100, 2)}%')
print(f'Chance do jogador "O" ganhar: {round((vitoriasO / jogosNaoEmpatados) * 100, 2)}%')

----- Jogos que não terminam empatados -----

Chance do jogador "X" ganhar: 66.86%
Chance do jogador "O" ganhar: 33.14%


Isto ocorre pelo fato de, como o jogador "X" é o primeiro jogar, ele limita inicialmente o *espaço amostral* das combinações possíveis para a vitória do jogador "O". Dessa forma, apesar da aleatoriedade da jogada do jogador "X" a cada rodada, as posições favoráveis ao jogador "O" na rodada seguinte é sempre menor. Isso fica claro quando analisamos, por exemplo, a primeira rodada:

- O jogador "X" preenche aleatoriamente uma casa e limita o tabuleiro a $8$ casas.

- Logo, a partir daí, todas as possibilidades de vitória do jogador "O" que utilizam a casa já preenchida por "X" são descartadas.

- Portanto, o jogador "O" não pode mais ganhar em uma das horizontais, uma das verticais e, em $\frac{5}{9}$ das vezes, uma das diagonais (se o "X" estiver em quaisquer $5$ das $9$ casas do tabuleiro as quais compõem uma das diagonais).

Assim, fica claro que o primeiro jogador, isto é, o jogador "X" tem mais chances de ganhar dentre os jogos que não terminam empatados.

#### c) Responda às mesmas perguntas dos itens a) e b), porém para um *jogo da velha tridimensional*: ou seja, em vez do jogo acontecer em uma grade bidimensional de tamanho $3 \times 3$, agora ele acontece em uma grade tridimensional de tamanho $3 \times 3 \times 3$! 🤯 Tome cuidado pois agora há diversas formas de um jogador ganhar uma partida! Além de ser através do completamento de alguma linha, coluna ou diagonal de qualquer "sub-tabuleiro" plano do cubo $3 \times 3 \times 3$ (existem 9 desses tabuleiros), o jogador poderá ainda completar as diagonais do cubo que unem vértices opostos (4 no total).

Tal como no jogo da velha comum, podemos encontrar a probabilidade do jogo não terminar em empate ao encontrar as chances de vitória do primeiro ou do segundo jogador. Logo, nosso código utiliza da mesma estrutura principal de simular uma grande quantidade de jogos e, assim, contabilizar o número de vitórias "X", vitórias "O" e empates.  

Porém, diferentemente do jogo comum, o jogo da velha tridimencional conta com bem mais combinações de vitória para os jogadores. Dessa forma, a cada rodada, verificamos se pelo menos uma das seguintes combinações ocorre nas casas do tabuleiro ao serem preenchidas com "X" ou "O" (considerando que os planos no jogo tridimencional são obtidos por secções horizontais entre os tabuleiros):

1. Uma das três linhas na horizontal em um mesmo plano completamente preenchida ($3 \times 3 \text{ planos} = 9 \text{ combinações}$);

2. Uma das três colunas na vertical em um mesmo plano completamente preenchida ($3 \times 3 \text{ planos} = 9 \text{ combinações}$);

3. Uma das nove colunas na vertical que corta os três planos completamente preenchida ($9$ combinações);

4. Uma das duas diagonais em um mesmo plano completamente preenchida ($2 \times 3 \text{ planos} = 6 \text{ combinações}$);

5. Uma das doze diagonais que cortam os três planos em uma mesma secção vertical "do cubo" completamente preenchida ($12$ combinações);

6. Uma das 4 diagonais que cortam os três planos indo de um vértice a outro "do cubo" completamente preenchida ($4$ combinações);


Assim, sendo inicialmente $49$ o total de combinações de vitória para cada jogador, podemos simular os jogos da velha com o seguinte código:

In [9]:
def verificaResultadoTridimencional(tabuleiro, rodada):
  # Casos de vitória "X"

  # Horizontal no mesmo plano
  if tabuleiro[0] == 1 and tabuleiro[1] == 1 and tabuleiro[2] == 1: return 1
  elif tabuleiro[3] == 1 and tabuleiro[4] == 1 and tabuleiro[5] == 1: return 1
  elif tabuleiro[6] == 1 and tabuleiro[7] == 1 and tabuleiro[8] == 1: return 1

  elif tabuleiro[9] == 1 and tabuleiro[10] == 1 and tabuleiro[11] == 1: return 1
  elif tabuleiro[12] == 1 and tabuleiro[13] == 1 and tabuleiro[14] == 1: return 1
  elif tabuleiro[15] == 1 and tabuleiro[16] == 1 and tabuleiro[17] == 1: return 1

  elif tabuleiro[18] == 1 and tabuleiro[19] == 1 and tabuleiro[20] == 1: return 1
  elif tabuleiro[21] == 1 and tabuleiro[22] == 1 and tabuleiro[23] == 1: return 1
  elif tabuleiro[24] == 1 and tabuleiro[25] == 1 and tabuleiro[26] == 1: return 1

  # Vertical no mesmo plano
  elif tabuleiro[0] == 1 and tabuleiro[3] == 1 and tabuleiro[6] == 1: return 1
  elif tabuleiro[1] == 1 and tabuleiro[4] == 1 and tabuleiro[7] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[5] == 1 and tabuleiro[8] == 1: return 1

  elif tabuleiro[9] == 1 and tabuleiro[12] == 1 and tabuleiro[15] == 1: return 1
  elif tabuleiro[10] == 1 and tabuleiro[13] == 1 and tabuleiro[16] == 1: return 1
  elif tabuleiro[11] == 1 and tabuleiro[14] == 1 and tabuleiro[17] == 1: return 1

  elif tabuleiro[18] == 1 and tabuleiro[21] == 1 and tabuleiro[24] == 1: return 1
  elif tabuleiro[19] == 1 and tabuleiro[22] == 1 and tabuleiro[25] == 1: return 1
  elif tabuleiro[20] == 1 and tabuleiro[23] == 1 and tabuleiro[26] == 1: return 1

  # Vertical entre planos
  elif tabuleiro[0] == 1 and tabuleiro[9] == 1 and tabuleiro[18] == 1: return 1
  elif tabuleiro[1] == 1 and tabuleiro[10] == 1 and tabuleiro[19] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[11] == 1 and tabuleiro[20] == 1: return 1

  elif tabuleiro[3] == 1 and tabuleiro[12] == 1 and tabuleiro[21] == 1: return 1
  elif tabuleiro[4] == 1 and tabuleiro[13] == 1 and tabuleiro[22] == 1: return 1
  elif tabuleiro[5] == 1 and tabuleiro[14] == 1 and tabuleiro[23] == 1: return 1

  elif tabuleiro[6] == 1 and tabuleiro[15] == 1 and tabuleiro[24] == 1: return 1
  elif tabuleiro[7] == 1 and tabuleiro[16] == 1 and tabuleiro[25] == 1: return 1
  elif tabuleiro[8] == 1 and tabuleiro[17] == 1 and tabuleiro[26] == 1: return 1

  # Diagonal no mesmo plano
  elif tabuleiro[0] == 1 and tabuleiro[4] == 1 and tabuleiro[8] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[4] == 1 and tabuleiro[6] == 1: return 1

  elif tabuleiro[9] == 1 and tabuleiro[13] == 1 and tabuleiro[17] == 1: return 1
  elif tabuleiro[11] == 1 and tabuleiro[13] == 1 and tabuleiro[15] == 1: return 1

  elif tabuleiro[18] == 1 and tabuleiro[22] == 1 and tabuleiro[26] == 1: return 1
  elif tabuleiro[20] == 1 and tabuleiro[22] == 1 and tabuleiro[24] == 1: return 1

  # Diagonal entre planos
  elif tabuleiro[0] == 1 and tabuleiro[10] == 1 and tabuleiro[20] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[10] == 1 and tabuleiro[18] == 1: return 1

  elif tabuleiro[3] == 1 and tabuleiro[13] == 1 and tabuleiro[23] == 1: return 1
  elif tabuleiro[5] == 1 and tabuleiro[13] == 1 and tabuleiro[21] == 1: return 1

  elif tabuleiro[6] == 1 and tabuleiro[16] == 1 and tabuleiro[26] == 1: return 1
  elif tabuleiro[8] == 1 and tabuleiro[16] == 1 and tabuleiro[24] == 1: return 1

  elif tabuleiro[0] == 1 and tabuleiro[12] == 1 and tabuleiro[24] == 1: return 1
  elif tabuleiro[6] == 1 and tabuleiro[12] == 1 and tabuleiro[18] == 1: return 1

  elif tabuleiro[1] == 1 and tabuleiro[13] == 1 and tabuleiro[25] == 1: return 1
  elif tabuleiro[7] == 1 and tabuleiro[13] == 1 and tabuleiro[19] == 1: return 1

  elif tabuleiro[2] == 1 and tabuleiro[14] == 1 and tabuleiro[26] == 1: return 1
  elif tabuleiro[8] == 1 and tabuleiro[14] == 1 and tabuleiro[20] == 1: return 1

  elif tabuleiro[0] == 1 and tabuleiro[13] == 1 and tabuleiro[26] == 1: return 1
  elif tabuleiro[2] == 1 and tabuleiro[13] == 1 and tabuleiro[24] == 1: return 1
  elif tabuleiro[6] == 1 and tabuleiro[13] == 1 and tabuleiro[20] == 1: return 1
  elif tabuleiro[8] == 1 and tabuleiro[13] == 1 and tabuleiro[18] == 1: return 1


  # Casos de vitória "O"

  # Horizontal no mesmo plano
  elif tabuleiro[0] == 2 and tabuleiro[1] == 2 and tabuleiro[2] == 2: return 2
  elif tabuleiro[3] == 2 and tabuleiro[4] == 2 and tabuleiro[5] == 2: return 2
  elif tabuleiro[6] == 2 and tabuleiro[7] == 2 and tabuleiro[8] == 2: return 2

  elif tabuleiro[9] == 2 and tabuleiro[10] == 2 and tabuleiro[11] == 2: return 2
  elif tabuleiro[12] == 2 and tabuleiro[13] == 2 and tabuleiro[14] == 2: return 2
  elif tabuleiro[15] == 2 and tabuleiro[16] == 2 and tabuleiro[17] == 2: return 2

  elif tabuleiro[18] == 2 and tabuleiro[19] == 2 and tabuleiro[20] == 2: return 2
  elif tabuleiro[21] == 2 and tabuleiro[22] == 2 and tabuleiro[23] == 2: return 2
  elif tabuleiro[24] == 2 and tabuleiro[25] == 2 and tabuleiro[26] == 2: return 2

  # Vertical no mesmo plano
  elif tabuleiro[0] == 2 and tabuleiro[3] == 2 and tabuleiro[6] == 2: return 2
  elif tabuleiro[1] == 2 and tabuleiro[4] == 2 and tabuleiro[7] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[5] == 2 and tabuleiro[8] == 2: return 2

  elif tabuleiro[9] == 2 and tabuleiro[12] == 2 and tabuleiro[15] == 2: return 2
  elif tabuleiro[10] == 2 and tabuleiro[13] == 2 and tabuleiro[16] == 2: return 2
  elif tabuleiro[11] == 2 and tabuleiro[14] == 2 and tabuleiro[17] == 2: return 2

  elif tabuleiro[18] == 2 and tabuleiro[21] == 2 and tabuleiro[24] == 2: return 2
  elif tabuleiro[19] == 2 and tabuleiro[22] == 2 and tabuleiro[25] == 2: return 2
  elif tabuleiro[20] == 2 and tabuleiro[23] == 2 and tabuleiro[26] == 2: return 2

  # Vertical entre planos
  elif tabuleiro[0] == 2 and tabuleiro[9] == 2 and tabuleiro[18] == 2: return 2
  elif tabuleiro[1] == 2 and tabuleiro[10] == 2 and tabuleiro[19] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[11] == 2 and tabuleiro[20] == 2: return 2

  elif tabuleiro[3] == 2 and tabuleiro[12] == 2 and tabuleiro[21] == 2: return 2
  elif tabuleiro[4] == 2 and tabuleiro[13] == 2 and tabuleiro[22] == 2: return 2
  elif tabuleiro[5] == 2 and tabuleiro[14] == 2 and tabuleiro[23] == 2: return 2

  elif tabuleiro[6] == 2 and tabuleiro[15] == 2 and tabuleiro[24] == 2: return 2
  elif tabuleiro[7] == 2 and tabuleiro[16] == 2 and tabuleiro[25] == 2: return 2
  elif tabuleiro[8] == 2 and tabuleiro[17] == 2 and tabuleiro[26] == 2: return 2

  # Diagonal no mesmo plano
  elif tabuleiro[0] == 2 and tabuleiro[4] == 2 and tabuleiro[8] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[4] == 2 and tabuleiro[6] == 2: return 2

  elif tabuleiro[9] == 2 and tabuleiro[13] == 2 and tabuleiro[17] == 2: return 2
  elif tabuleiro[11] == 2 and tabuleiro[13] == 2 and tabuleiro[15] == 2: return 2

  elif tabuleiro[18] == 2 and tabuleiro[22] == 2 and tabuleiro[26] == 2: return 2
  elif tabuleiro[20] == 2 and tabuleiro[22] == 2 and tabuleiro[24] == 2: return 2

  # Diagonal entre planos
  elif tabuleiro[0] == 2 and tabuleiro[10] == 2 and tabuleiro[20] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[10] == 2 and tabuleiro[18] == 2: return 2

  elif tabuleiro[3] == 2 and tabuleiro[13] == 2 and tabuleiro[23] == 2: return 2
  elif tabuleiro[5] == 2 and tabuleiro[13] == 2 and tabuleiro[21] == 2: return 2

  elif tabuleiro[6] == 2 and tabuleiro[16] == 2 and tabuleiro[26] == 2: return 2
  elif tabuleiro[8] == 2 and tabuleiro[16] == 2 and tabuleiro[24] == 2: return 2

  elif tabuleiro[0] == 2 and tabuleiro[12] == 2 and tabuleiro[24] == 2: return 2
  elif tabuleiro[6] == 2 and tabuleiro[12] == 2 and tabuleiro[18] == 2: return 2

  elif tabuleiro[1] == 2 and tabuleiro[13] == 2 and tabuleiro[25] == 2: return 2
  elif tabuleiro[7] == 2 and tabuleiro[13] == 2 and tabuleiro[19] == 2: return 2

  elif tabuleiro[2] == 2 and tabuleiro[14] == 2 and tabuleiro[26] == 2: return 2
  elif tabuleiro[8] == 2 and tabuleiro[14] == 2 and tabuleiro[20] == 2: return 2

  elif tabuleiro[0] == 2 and tabuleiro[13] == 2 and tabuleiro[26] == 2: return 2
  elif tabuleiro[2] == 2 and tabuleiro[13] == 2 and tabuleiro[24] == 2: return 2
  elif tabuleiro[6] == 2 and tabuleiro[13] == 2 and tabuleiro[20] == 2: return 2
  elif tabuleiro[8] == 2 and tabuleiro[13] == 2 and tabuleiro[18] == 2: return 2

  else:
    # Verifica se é a última rodada e determina o empate
    if rodada == 26: return 0
    else: return -1


# Simulando os jogos
numJogos = 100000
vitoriasX = 0.
vitoriasO = 0.
empates = 0.

for i in range(numJogos):
  resultado = jogoDaVelha(27)
  if resultado == 0: empates += 1
  elif resultado == 1: vitoriasX += 1
  elif resultado == 2: vitoriasO += 1

# Imprimindo resultados
print(f'Vitórias X: {round((vitoriasX / numJogos) * 100, 2)}%')
print(f'Vitórias O: {round((vitoriasO / numJogos) * 100, 2)}%')
print(f'Empates: {round((empates / numJogos) * 100, 2)}%')

jogosNaoEmpatados = vitoriasX + vitoriasO

print('\n----- Jogos que não terminam empatados -----\n')
print(f'Chance do jogador "X" ganhar: {round((vitoriasX / jogosNaoEmpatados) * 100, 2)}%')
print(f'Chance do jogador "O" ganhar: {round((vitoriasO / jogosNaoEmpatados) * 100, 2)}%')

Vitórias X: 56.37%
Vitórias O: 43.63%
Empates: 0.0%

----- Jogos que não terminam empatados -----

Chance do jogador "X" ganhar: 56.37%
Chance do jogador "O" ganhar: 43.63%


Portanto, dadas as probabilidades acima considerando uma simulação de $100000$ jogos, podemos concluir que a chance de um jogo não terminar em empate é de:

$\text{vitóriasX + vitoriasO} = 56,73\% +  43,27\% = 100\%$

Como vemos, a chance de haver um empate em um jogo da velha tridimencional cai para $0\%$, o que se dá justamente pela enorme quantidade de combinações de vitória para cada jogador ($\frac{49}{8} = 6,125$ vezes maior que no jogo tradicional), tornando praticamente impossível que uma partida termine sem que nenhum dos jogadores a ganhe.

Além disso, podemos ver que, assim como no jogo da velha comum, a probabilidade do primeiro jogador, isto é, o jogador "X" ganhar, dentre os jogos que não terminam empatados, é maior que a do jogador "O".

No entanto, como o *espaço amostral* de jogos não empatados é maior, observamos uma leve redução dessa probabilidade do jogador "X" ganhar em relação ao visto no jogo da velha tradicional. Isso se dá pelo fato de, apesar do jogador "X" ser o primeiro jogar e, portanto, limitar inicialmente o *espaço amostral* das combinações possíveis para a vitória do jogador "O", no jogo tridimencional ele está limitando **uma parte menor** desse espaço amostral. Podemos observar isso ao analisar, por exemplo, a primeira rodada:

- O jogador "X" preenche aleatoriamente uma casa e limita o tabuleiro a $26$ casas.

- Logo, a partir daí, todas as possibilidades de vitória do jogador "O" que utilizam a casa já preenchida por "X" são descartadas, porém ainda há um número considerável de combinações que possibitam essa vitória.

- Dessa forma, o jogador "O" não pode mais ganhar em uma das horizontais de um dos planos, uma das verticais de um dos planos, uma das verticais que corta os três planos e, em $\frac{19}{27}$ das vezes, uma das diagonais (se o "X" estiver em quaisquer $19$ das $27$ casas do tabuleiro as quais compõem uma das diagonais, seja de um mesmo plano ou uma que corta os planos).

Vemos então que, no pior dos casos do jogo tridimencional (o caso em que o "X" cai na casa central "do cubo": `tabuleiro[13]`), ainda haverá $49 - 13 = 36$ combinações de vitória, enquanto que no jogo comum, no pior caso (o caso em que o "X" cai na casa central do tabuleiro: `tabuleiro[4]`) há $8 - 4 = 4$ combinações de vitória. Quando comparamos relativamente as combinações de vitória após a primeira rodada considerando o pior caso de posição do "X", vemos que no jogo tridimencional temos $\frac{36}{49} ≈ 73\%$ de chance de vitória do jogador "O", enquanto que no jogo comum temos apenas $\frac{4}{8} = 50\%$ de chance de vitória do jogador "O".

Assim, fica claro que o primeiro jogador, isto é, o jogador "X" continua tendo mais chances de ganhar em um jogo tridimencional dentre os jogos que não terminam empatados, porém essa chance é um pouco menor se comparada ao jogo da velha tradicional.